walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,206 @@
+
+    // If the meeting duration is over 24 hours, then return no available timeRange.
+    if (meetingDurationMinutes > TimeRange.WHOLE_DAY.duration()) {
+      return Collections.emptyList();
+    }
+
+    // If no Attendees are passed in, return a timeRange containing the whole day.
+    if (request.getAttendees().isEmpty() && request.getOptionalAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Store unavailable TimeRanges for mandatory and optional attendees separately.
+    List<TimeRange> unavailableMandatoryTimeRanges =
+        getUnavailableTimeRanges(events, mandatoryAttendees);
+    List<TimeRange> unavailableOptionalTimeRanges =
+        getUnavailableTimeRanges(events, optionalAttendees);
+
+    // Get available TimeRanges from unavailable TimeRanges.
+    List<TimeRange> availableMandatoryTimeRanges =
+        getAvailableTimeRanges(unavailableMandatoryTimeRanges, meetingDurationMinutes);
+    List<TimeRange> availableOptionalTimeRanges =
+        getAvailableTimeRanges(unavailableOptionalTimeRanges, meetingDurationMinutes);
+
+    // If no mandatory attendees are passed, return the available TimeRanges for optional attendees
+    // else if no optional attendees are passed, or no TimeRanges are available for optional
+    // attendees then return available TimeRanges for mandatory attendees, else merge the available
+    // TimeRanges for mandatory attendees with available TimeRanges for optional employees.
+    if (mandatoryAttendees.isEmpty()) {
+      return availableOptionalTimeRanges;
+    } else if (optionalAttendees.isEmpty() || availableOptionalTimeRanges.isEmpty()) {
+      return availableMandatoryTimeRanges;
+    } else {
+      return mergeAvailableTimeRanges(
+          availableMandatoryTimeRanges, availableOptionalTimeRanges, meetingDurationMinutes);
+    }
+  }
+
+  /** Returns a List of TimeRanges that are unavailable in as little TimeRanges as possible. */
+  private List<TimeRange> getUnavailableTimeRanges(
+      Collection<Event> events, Collection<String> attendees) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    // If an event's attendee list contains a names of an attendee that needs to attend
+    // (from Collection<String> attendees, then we add the event's TimeRange to a list.
+    events.forEach(
+        event -> {
+          Collection<String> eventAttendees = event.getAttendees();
+
+          for (String attendee : eventAttendees) {
+            if (!attendees.contains(attendee)) {
+              continue;
+            }
+
+            TimeRange currentEventTimeRange = event.getWhen();
+
+            // If the currentEventTimeRange isn't merged into a TimeRange within the
+            // unavailableTimeRanges list, then add the currentEventTimeRange to the list.
+            if (!mergeOnOverlap(unavailableTimeRanges, currentEventTimeRange)) {
+              unavailableTimeRanges.add(currentEventTimeRange);
```

Comment:
You're passing `unavailableTimeRanges` into `mergeOnOverlap(...)` anyway, so why not just do this step inside that method? Right now you're splitting up the logic of how `unavailableTimeRanges` is mutated into two places and it would be simpler to follow later if it was all together in one spot.

Reply:
Done.